linode/terraform-provider-linode
# new: Add support for batch API requests

## 📝 Description

This change adds support for concisely running non-sequential API requests concurrently. This can considerably speed up plans for configurations that contain a large number of resources and for users with high-latency connections.

This pull request only parallelizes GET requests in instance-related resources, we can prioritize moving other resources over in the future.

The performance gains may seem small (~1 second) on a minimal configuration or on a low-latency connection, but the performance improvement should scale considerably with less ideal conditions.

## ✔️ How to Test

The following test steps assume you have pulled down this PR locally.

### Integration Testing

```
make int-test PKG_NAME=linode/instance
make int-test PKG_NAME=linode/instanceconfig
```

### Unit Testing

```
make unit-test
```

### Manual Testing

1. In a Terraform provider sandbox environment (e.g. dx-devenv), apply the following configuration:

```terraform
resource "linode_instance_config" "my-config" {
  linode_id = linode_instance.my-instance.id
  label = "my-config"

  device {
    device_name = "sda"
    disk_id = linode_instance_disk.boot.id
  }

  booted = true
}

resource "linode_instance_disk" "boot" {
  label = "boot"
  linode_id = linode_instance.my-instance.id
  size = linode_instance.my-instance.specs.0.disk

  image = "linode/ubuntu22.04"
  root_pass = "myc00lpa$$w0rd!!!!!!"
}

resource "linode_instance" "my-instance" {
  label = "my-instance"
  type = "g6-nanode-1"
  region = "us-mia"
}
```

2. Ensure the configuration applies successfully.
